### PR TITLE
Instrument code for logging compilation time.

### PR DESCRIPTION
**Pull request by TejashShah**
Only in ~HIP backend and~ DEV build type, one could gather individual kernel compilation time _and total compilation time for all kernels_

### CODE REVIEW

**Comment by atamazov** [comment]
I think we'll generalize this. I need some free time to tell more. How urgent is this?

File: src/solver.cpp [hunk 1]
```diff
@@ -80,9 +80,21 @@ void PrecompileSolutions(const Handle& h, const std::vector<ConvSolution>& sols)
...
+
     // Precompile the kernels in parallel, but dont add them to the cache
     std::vector<Program> programs = PrecompileKernels(h, kernels);
 
+#if MIOPEN_BUILD_DEV
+    auto compileEnd = std::chrono::steady_clock::now();
+    auto timems = std::chrono::duration_cast<std::chrono::duration<float, std::milli>>(compileEnd -
+                                                                                       compileStart)
+                      .count();
+    MIOPEN_LOG_I2("ConvSolution Compile Time: " << timems << " ms");
```
[the remark this answers is not in the document]
Comment (TejashShah) [reply src/solver.cpp:95]: Will do

File: src/hip/handlehip.cpp [hunk 2]
```diff
@@ -371,9 +371,21 @@ Program Handle::LoadProgram(const std::string& program_name,
         this->GetDeviceName(), this->GetMaxComputeUnits(), program_name, params, is_kernel_str);
     if(hsaco.empty())
     {
+
+#if MIOPEN_BUILD_DEV
```
[the remark this answers is not in the document]
Comment (TejashShah) [reply src/hip/handlehip.cpp:375]: Sounds good.

File: src/solver.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (atamazov) [reply src/solver.cpp:95]: And better move this into PrecompileKernels:
```C++
std::vector<Program> PrecompileKernels(const Handle& h, const std::vector<KernelInfo>& kernels)
{
    CompileTimer ct;
...
    ct.print("PrecompileKernels");
    return programs;
}
```

[the remark this answers is not in the document]
Comment (atamazov) [reply src/solver.cpp:95]: ...so each invocation of `PrecompileKernels()` will print.

File: src/hip/handlehip.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TejashShah) [reply src/hip/handlehip.cpp:375]: I think, we should keep both compile_timer and timer classes in a single timer.hpp. Creating 2 files for a single concept of Timer seems unnecessary.

[the remark this answers is not in the document]
Comment (atamazov) [reply src/hip/handlehip.cpp:375]: This is very special timer/printer intended to measure/print only build times. And it should include logger.hpp...  I do not see any harm in adding a small specialized module for it. But it's up to you.

File: src/solver.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (TejashShah) [reply src/solver.cpp]: done

File: src/hip/handlehip.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TejashShah) [reply src/hip/handlehip.cpp]: done in a single file

File: src/solver.cpp [hunk 3]
```diff
@@ -54,13 +55,17 @@ std::ostream& operator<<(std::ostream& os, const KernelInfo& k)
...
+	
     std::vector<Program> programs(kernels.size());
     par_for(kernels.size(),
             max_threads{Value(MIOPEN_COMPILE_PARALLEL_LEVEL{}, 20)},
             [&](auto i) {
                 const KernelInfo& k = kernels[i];
                 programs[i]         = h.LoadProgram(k.kernel_file, k.comp_options, false, "");
             });
+
+    ct.Log("Kernel");
```
Comment (atamazov) [inline src/solver.cpp:68]:
```suggestion
    ct.Log("PrecompileKernels");
```

File: src/include/miopen/timer.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*
+*******************************************************************************/
+#ifndef GUARD_MIOPEN_TIMER_HPP_
+#define GUARD_MIOPEN_TIMER_HPP_
+
+#include <chrono>
+
+#include <miopen/logger.hpp>
```
Comment (atamazov) [inline src/include/miopen/timer.hpp:31]: Local headers first, please

File: src/include/miopen/timer.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/*******************************************************************************
+*
+* MIT License
+*
+* Copyright (c) 2019 Advanced Micro Devices, Inc.
```
Comment (atamazov) [inline src/include/miopen/timer.hpp:5]: This is my favorite ;)

File: src/include/miopen/timer.hpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    CompileTimer()
+    {
+#if MIOPEN_BUILD_DEV
+        timer.start();
+#endif
+    }
+    void Log(const std::string& s1, const std::string& s2 = {})
+    {
+#if MIOPEN_BUILD_DEV
+        MIOPEN_LOG_I(
```
Comment (atamazov) [inline src/include/miopen/timer.hpp:68]: Info2, please

File: src/hip/handlehip.cpp [hunk 7]
```diff
@@ -371,8 +372,11 @@ Program Handle::LoadProgram(const std::string& program_name,
         this->GetDeviceName(), this->GetMaxComputeUnits(), program_name, params, is_kernel_str);
     if(hsaco.empty())
     {
+
```
Comment (atamazov) [inline src/hip/handlehip.cpp:375]: Please remove this empty line

**Comment by atamazov** [comment]
@TejashShah I've updated PR description

**Comment by atamazov** [comment]
Please merge recent develop into this PR or it will stuck at COMGR testing phase.

**Pull request merged**